src/storage/pubsub/listener.c

```diff
@@ -0,0 +1,207 @@
+#include "hashtable.h"
+#include "listener.h"
+
+#include <cc_mm.h>
+#include <cc_queue.h>
+#include <hash/cc_lookup3.h>
+
+#include <sysexits.h>
+
+SLIST_HEAD(listener_slh, listener);
+
+struct listener_ht {
+    struct listener_slh *table;
+    uint32_t nlistener;
+    uint32_t hash_power;
+};
+
+static struct listener_ht hashtable;
+static struct listener_ht *ht = &hashtable;
+
+
+static struct listener_slh *
+_get_bucket(const struct buf_sock *s)
+{
+    /* use the _address_ of the channel to hash */
+    uint32_t hval = hash_lookup3((char *)s, sizeof(struct buf_sock *), 0);
+    return &(ht->table[hval & HASHMASK(ht->hash_power)]);
+}
+
+struct listener *
+listener_get(const struct buf_sock *s)
+{
+    struct listener_slh *bucket;
+    struct listener *l;
+
+    bucket = _get_bucket(s);
+    for (l = SLIST_FIRST(bucket); l != NULL; l = SLIST_NEXT(l, l_sle)) {
+        if (l->s == s) {
+            return l;
+        }
+    }
+
+    return NULL;
+}
+
+void
+listener_put(const struct listener *l)
+{
+    struct listener_slh *bucket;
+
+    ASSERT(listener_get(l->s) == NULL);
+
+    bucket = _get_bucket(l->s);
+    SLIST_INSERT_HEAD(bucket, (struct listener *)l, l_sle);
+
+    ht->nlistener++;
+}
+
+void
+listener_delete(const struct buf_sock *s)
+{
+    struct listener_slh *bucket;
+    struct listener *l, *prev;
+
+    bucket = _get_bucket(s);
+    for (prev = NULL, l = SLIST_FIRST(bucket); l != NULL;
+        prev = l, l = SLIST_NEXT(l, l_sle)) {
+        if (l->s == s) {
+            break;
+        }
+    }
+
+    if (prev == NULL) {
+        SLIST_REMOVE_HEAD(bucket, l_sle);
+    } else {
+        SLIST_REMOVE_AFTER(prev, l_sle);
```

Comment:
You are right, the way I'm using it now assumes `l != NULL`, but I might as well allow passing in an invalid listener to simplify the deletion logic.

Eventually I want to add pooling to listener (and topic) to avoid calling malloc & free all the time, I will later think about if I want to allow people to create/destroy listener objects directly.